ouroboros-consensus/src/Ouroboros/Consensus/ChainSyncClient.hs

```diff
@@ -767,6 +763,28 @@ rejectInvalidBlocks tracer registry getIsInvalidBlock getCandidate =
       traceWith tracer $ TraceException ex
       throwM ex
 
+rollback
+  :: ( HasHeader (Header blk)
+     , OuroborosTag (BlockProtocol blk)
+     )
+  => NodeConfig (BlockProtocol blk)
+  -> AnchoredFragment (Header blk)
+  -> ChainState (BlockProtocol blk)
+  -> Point blk
+  -> Maybe (AnchoredFragment (Header blk), ChainState (BlockProtocol blk))
+rollback cfg ourFrag ourChainState intersection = do
+    let i = castPoint intersection
+    frag' <- AF.rollback i ourFrag
+
+    let slot = pointSlot i
+    let isAccompaniedEBB = case AF.successorBlock i ourFrag of
+          Nothing  -> False
+            -- intersection might be an EBB, but not an accompanied one
+          Just hdr -> At (blockSlot hdr) == slot
+    cs' <- rewindChainState cfg ourChainState slot isAccompaniedEBB
+
+    pure (frag', cs')
+
```

Comment:
I'm thinking about the alternative solution you proposed when we first ran into the EBB+PBFT.ChainState problem:

* Make the ChainSyncClient EBB-aware, i.e. let it take a `Header blk -> IsEBB` function.
* In `rollback`: if `isEBB i == IsEBB`, then look at the predecessor (I'll write `AF.predecessorBlock`) and use its `SlotNo` to rewind the chain state. In case of the Genesis EBB, rewind to `Origin`. *Problem:* what if the rollback point is the anchor point of the fragment? Then we can't look at its predecessor. Maybe it's fine to throw `InvalidRollBack` (or a new exception constructor?) in that case. I'll think some more about this.
* In `PBFT.ChainState`: remove all traces from EBBs (and make Edsko happy again :slightly_smiling_face: ). (And avoid #1356! :smile:) 
* No need to touch the signature of `rewindChainState`.
* Modify the decoder of `PBFT.ChainState` to ignore the EBB field and increase the version number.

Am I forgetting something? I'm thinking now I made the wrong decision the first time :grin:. The main downside for me is that the ChainSyncClient is now EBB-aware, which feels a bit dirty, as EBBs are specific to one consensus protocol.

Reply:
Hmm.

> The main downside for me is that the ChainSyncClient is now EBB-aware, which feels a bit dirty, as EBBs are specific to one consensus protocol.

I agree: it seems better for the PBFT ChainState to be complicated by EBBs, since PBFT actually involves EBBs.

Moreover, what's the main upside? The one listed is

> No need to touch the signature of `rewindChainState`.

but that's not particularly compelling? In this workaround, I've added an "is EBB" argument to `rewindChainState`. We could instead change its `SlotNo` argument to a `Point` and have the chain state retain hashes. That's also be viable, and wouldn't explicitly involve EBBs (it'd only explicitly allow for multiple blocks in a slot, which we (and the PBFT code) know currently only happens with EBBs.)

Reply:
By the way, the `ebbs` field could be abstracted out of the `PBFT` chain state, if we wanted to recover the "pure PBFT" chain state.

Reply:
> Hmm.
> 
> > The main downside for me is that the ChainSyncClient is now EBB-aware, which feels a bit dirty, as EBBs are specific to one consensus protocol.
> 
> I agree: it seems better for the PBFT ChainState to be complicated by EBBs, since PBFT actually involves EBBs.
> 
> Moreover, what's the main upside? The one listed is
> 
> > No need to touch the signature of `rewindChainState`.

True. My first draft of that comment included "Much simpler .. blah blah", but then I added the "*Problem*" part and removed it.

> but that's not particularly compelling? In this workaround, I've added an "is EBB" argument to `rewindChainState`. We could instead change its `SlotNo` argument to a `Point` and have the chain state retain hashes. That'd also be viable and wouldn't explicitly involve EBBs (it'd only explicitly allow for multiple blocks in a slot, which we (and the PBFT code) know currently only happens with EBBs.)

I like that idea :+1: Let me think about it.

Reply:
> > [..] In this workaround, I've added an "is EBB" argument to `rewindChainState`. We could instead change its `SlotNo` argument to a `Point` and have the chain state retain hashes. That'd also be viable and wouldn't explicitly involve EBBs (it'd only explicitly allow for multiple blocks in a slot, which we (and the PBFT code) know currently only happens with EBBs.)
> 
> I like that idea +1 Let me think about it.

Concretely, this means that `PBFT.ChainState` will track the hash of each applied EBB, so we can intercept a `rewindChainState` call to an EBB. The hash is ignored for regular blocks. Hmm, it feels a bit wrong to pass in the hash if we're not using it to check whether it matches a previously applied header. Not that I'm arguing we should, as that would mean that we'd have to store the hashes too, requiring even more bytes, and breaking binary compatibility. It is less ad-hoc than the `Bool` argument you added to `rewindChainState`.

Reply:
If you're ready to commit to that route, would you update Issue 1352 (which blocks this PR)? And would you want to open that PR yourself or shall I?

Reply:
It's not ideal, but yeah, let's go for it. I'll leave the PR up to you :slightly_smiling_face:

Reply:
Resolution: implemented.